File: leaf_mobile/lib/utils/page_cache.dart
// lib/utils/page_cache.dart
//
// ── Cache-first na pattern para sa member app pages — instant na
// ipinapakita ang huling nakitang datos habang tahimik na nagre-
// refresh sa likod, imbes na palaging nagpapakita ng loading spinner
// tuwing may bagong pag-navigate. Katumbas ito ng web's pageCache.js.
//
// MAHALAGA: static IN-MEMORY na Map ito (hindi SharedPreferences) —
// sinasadya, dahil ang SharedPreferences ay ASYNCHRONOUS mag-basa sa
// Flutter (may async gap bago available ang laman, kaya may mag-
// fli-flicker pa rin). Ang static in-memory Map ay SYNCHRONOUS —
// available agad sa unang frame, walang async gap, katumbas ng
// sessionStorage sa web (nabubuhay habang bukas ang app session,
// nawawala kapag talagang isinara/kina-kill ang app).
//
// Naka-scope ang bawat cache entry sa account (scopeKey) — kaya kahit
// magpalit ng account sa parehong app session, hindi makikita ng
// bagong account ang cached data ng nauna.

class PageCache {
  static final Map<String, dynamic> _store = {};

  static String _key(String pageKey, String? scopeKey) => '${pageKey}_${scopeKey ?? "anon"}';

  static T? get<T>(String pageKey, String? scopeKey) {
    final v = _store[_key(pageKey, scopeKey)];
    return v is T ? v : null;
  }

  static void set(String pageKey, String? scopeKey, dynamic data) {
    _store[_key(pageKey, scopeKey)] = data;
  }

  // ── Ginagamit sa logout — para talagang malinis lahat ng naka-cache
  // na datos ng dating account. ────────────────────────────────────
  static void clearAll() {
    _store.clear();
  }
}

File: frontend/leaf_app/src/utils/pageCache.js
// src/utils/pageCache.js
// ── BAGO: cache-first na pattern para sa member portal pages — instant
// na ipinapakita ang huling nakitang datos habang tahimik na nagre-
// refresh sa likod, imbes na palaging nagpapakita ng "Loading..."
// tuwing may bagong pag-navigate. Ginagamit ng Dashboard, My Loans,
// My Savings, Announcements, My Profile, at Apply for Loan.
//
// MAHALAGA: naka-scope ang bawat cache key sa member ID (hindi lang
// sa pangalan ng page) — kaya kahit magpalit ng account sa parehong
// tab/session (hal. admin → member, o member A → member B), hindi
// makikita ng bagong account ang lumang cached data ng nauna. Kung
// walang alam na member ID pa (bago pa lang mag-load), i-treat na
// lang bilang "anon" — hindi ito magiging risk dahil isang beses lang
// gagamitin bago pa dumating ang totoong member ID.

const PREFIX = "leaf_pgcache_";

export function getPageCache(pageKey, memberId) {
  try {
    const raw = sessionStorage.getItem(`${PREFIX}${pageKey}_${memberId ?? "anon"}`);
    if (!raw) return null;
    const parsed = JSON.parse(raw);
    return parsed ?? null;
  } catch {
    return null;
  }
}

export function savePageCache(pageKey, memberId, data) {
  try {
    sessionStorage.setItem(`${PREFIX}${pageKey}_${memberId ?? "anon"}`, JSON.stringify(data));
  } catch { /* ignore — hindi kritikal, cache lang ito */ }
}

// ── Ginagamit sa logout — para talagang malinis lahat ng naka-cache
// na datos ng dating account, hindi lang yung "anon" placeholder. ────
export function clearAllPageCaches() {
  try {
    const keys = [];
    for (let i = 0; i < sessionStorage.length; i++) {
      const k = sessionStorage.key(i);
      if (k && k.startsWith(PREFIX)) keys.push(k);
    }
    keys.forEach(k => sessionStorage.removeItem(k));
  } catch { /* ignore */ }
}